### PR TITLE
Add a monitor for ElasticDL job.

### PR DESCRIPTION
**Pull request by workingloong**
Add 3 methods into Client to get logs of the master, workers, and ps.

### CODE REVIEW

**Review by terrytangyuan — COMMENTED** [review]
Where are these being used?

File: elasticdl/python/common/k8s_job_monitor.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+            else:
+                print(pod.status.phase)
+                time.sleep(30)
+
+        self.core_api.delete_namespaced_pod(
+            name=self.pod_name,
+            namespace=self.namespace,
+            body=client.V1DeleteOptions(grace_period_seconds=0),
+        )
+        print("End")
```
Comment (terrytangyuan) [inline elasticdl/python/common/k8s_job_monitor.py]: Remove unnecessary print

File: elasticdl/python/common/k8s_job_monitor.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+            worker_pod_name = self.client.get_worker_pod_name(i)
+            if worker_pod is None:
+                print("Worker {} Not Found".format(worker_pod_name))
+            elif worker_pod.status.phase == PodStatus.FAILED:
+                logger.error(
+                    "Worker {} {}".format(
+                        worker_pod_name, worker_pod.status.phase
+                    )
+                )
+                print("{} log :".format(worker_pod_name))
```
Comment (terrytangyuan) [inline elasticdl/python/common/k8s_job_monitor.py]: Change print to logging

File: elasticdl/python/common/k8s_job_monitor.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+            print("Master status: {}".format(master_pod.status.phase))
+            if master_pod.status.phase == PodStatus.SUCCEEDED:
+                job_success = True
+                break
+            elif master_pod.status.phase in [
+                PodStatus.PENDING,
+                PodStatus.CREATING,
+            ]:
+                time.sleep(120)
```
Comment (terrytangyuan) [inline elasticdl/python/common/k8s_job_monitor.py]: Are you sure? This is very long time

File: elasticdl/python/common/k8s_job_monitor.py (hunk 2 above)
Comment (terrytangyuan) [reply elasticdl/python/common/k8s_job_monitor.py]: Other places too

File: elasticdl/python/common/k8s_job_monitor.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+            use_kube_config: If true, load the cluster config from
+                ~/.kube/config. Otherwise, if it's in a process running in
+                a K8S environment, it loads the incluster config.
+        """
+        self.namespace = namespace
+        self.pod_name = pod_name
+        self.core_api = client.CoreV1Api()
+
+        try:
+            if use_kube_config:
```
Comment (terrytangyuan) [inline elasticdl/python/common/k8s_job_monitor.py]: Can we reuse existing code?

File: elasticdl/python/common/k8s_job_monitor.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+                time.sleep(30)
+
+        self.core_api.delete_namespaced_pod(
+            name=self.pod_name,
+            namespace=self.namespace,
+            body=client.V1DeleteOptions(grace_period_seconds=0),
+        )
+        print("End")
+
+    def get_pod(self):
```
Comment (terrytangyuan) [inline elasticdl/python/common/k8s_job_monitor.py]: Move these methods to k8s_client?

File: elasticdl/python/common/k8s_job_monitor.py (hunk 4 above)
Comment (workingloong) [reply elasticdl/python/common/k8s_job_monitor.py]: I move the `get_pod` and `get_pod_log` to `k8s_client`. Then we don't need those codes.

File: elasticdl/python/common/k8s_job_monitor.py (hunk 5 above)
Comment (workingloong) [reply elasticdl/python/common/k8s_job_monitor.py]: Done

File: elasticdl/python/common/k8s_job_monitor.py (hunk 3 above)
Comment (workingloong) [reply elasticdl/python/common/k8s_job_monitor.py]: The job execution job varies from minutes to hours. And I have reduced it to 60s. It is difficult to set the interval.

File: elasticdl/python/common/k8s_job_monitor.py (hunk 2 above)
Comment (workingloong) [reply elasticdl/python/common/k8s_job_monitor.py]: Yeah! Done

File: elasticdl/python/common/k8s_job_monitor.py (hunk 1 above)
Comment (workingloong) [reply elasticdl/python/common/k8s_job_monitor.py]: Done

File: elasticdl/python/common/k8s_job_monitor.py (hunk 3 above)
Comment (terrytangyuan) [reply elasticdl/python/common/k8s_job_monitor.py]: Sure but this is only the pod pending/creating time instead of the job execution time

Comment (workingloong) [reply elasticdl/python/common/k8s_job_monitor.py]: Oh, I have mixed the two intervals. May 10s be enough?

Comment (terrytangyuan) [reply elasticdl/python/common/k8s_job_monitor.py]: This depends on the cluster.

Comment (workingloong) [reply elasticdl/python/common/k8s_job_monitor.py]: Set a default value and the user can set it in arguments?

File: elasticdl/python/common/k8s_client.py [hunk 6]
```diff
@@ -384,29 +376,22 @@ def create_ps(self, **kargs):
...
 
     def delete_worker(self, worker_id):
-        self.client.delete_namespaced_pod(
-            self.get_worker_pod_name(worker_id),
-            self.namespace,
-            body=client.V1DeleteOptions(grace_period_seconds=0),
-        )
+        self.delete_pod(self.get_worker_pod_name(worker_id))
 
     def delete_embedding_service(self, embedding_service_id):
```
Comment (skydoorkai) [inline elasticdl/python/common/k8s_client.py]: There is no embedding service pod now. All related methods can be deleted.

Comment (workingloong) [reply elasticdl/python/common/k8s_client.py]: OK

File: elasticdl/python/common/k8s_job_monitor.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+            logger.info("Pod Status : %s" % pod.status.phase)
+            if pod.status.phase == PodStatus.SUCCEEDED:
+                pod_succeeded = True
+                break
+            elif pod.status.phase == PodStatus.FAILED:
+                logger.info(self.client.get_pod_log(self.pod_name))
+                break
+            else:
+                time.sleep(30)
+        self.client.delete_pod(self.pod_name)
```
Comment (skydoorkai) [inline elasticdl/python/common/k8s_job_monitor.py]: This method does not only monitor the pod status, but it also deletes the pod. May rename the method name to reflect this.

File: elasticdl/python/common/k8s_job_monitor.py [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+            ps_pod = self.client.get_ps_pod(i)
+            ps_pod_name = self.client.get_ps_pod_name(i)
+            if ps_pod is None:
+                logger.error("PS {} Not Found".format(ps_pod_name))
+            elif ps_pod.status.phase == PodStatus.FAILED:
+                logger.error(
+                    "PS {} {}".format(ps_pod_name, ps_pod.status.phase)
+                )
+
+    def show_master_increment_log(self, old_log):
```
Comment (skydoorkai) [inline elasticdl/python/common/k8s_job_monitor.py]: Add a comment to explain this method.

File: elasticdl/python/common/k8s_job_monitor.py (hunk 7 above)
Comment (workingloong) [reply elasticdl/python/common/k8s_job_monitor.py]: I have split the method into two methods `monitor_status` and `delete_pod`

**Comment by workingloong** [comment]
Close it because the branch-0.2 will be abandoned. The code of the PR will be merged to develop branch in #1799

**Pull request closed without merging**